pkg/errgroup/errgroup.go

```diff
@@ -26,19 +26,28 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
+const DefaultErrgroupConcurrency = 5
+
 type Group struct {
 	*errgroup.Group
-	logger logger.Logger
-	ctx    context.Context
+	logger        logger.Logger
+	semaphoreChan chan bool
+	ctx           context.Context
 }
 
-func WithContext(ctx context.Context, loggerInstance logger.Logger) (*Group, context.Context) {
+func WithContext(ctx context.Context, loggerInstance logger.Logger, concurrency int) (*Group, context.Context) {
 	newBaseErrgroup, errgroupCtx := errgroup.WithContext(ctx)
 
+	if concurrency <= 0 {
```

Comment:
We to give the developer the ability to overriding the semaphore. either by dependency injection of `concurrency=0` which means, no semaphore or having a different a constructor (aka function that create the object)

I'd go with:

```go
func WithContext(ctx context.Context, loggerInstance logger.Logger) (*Group, context.Context) {
...
}
```

vs

```go
func WithContextSemaphore(ctx context.Context, loggerInstance logger.Logger, concurrency int) (*Group, context.Context) {
...
}
```

both will implement the same thing, where `WithContext` will call `WithContextSemaphore` with `concurrency=0`.

That way, the existing calls will not perform (and change their behavior) concurrency based on semaphore
and you will pick up exactly where you _do_ want to perform semaphore.

Reply:
So when `concurreny=0` there won't be use of semaphore at all?

Also, should the only explicit call to `WithContextSemaphore`, as of now, be in `synchronizer.go` (where we sync projects from leader to project's CRD)?

Reply:
> So when `concurreny=0` there won't be use of semaphore at all?
>  

correct

> Also, should the only explicit call to `WithContextSemaphore`, as of now, be in `synchronizer.go` (where we sync projects from leader to project's CRD)?
>  

correct